{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## IMPORT\n",
    "import sys\n",
    "import os\n",
    "import app_paths\n",
    "import config\n",
    "import imp\n",
    "import f\n",
    "import f_nav\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## LOCAL\n",
    "imp.reload(config)\n",
    "giab_vcf_dict=config.giab_vcf_dict\n",
    "giab_bed_dict=config.giab_bed_dict\n",
    "ref_dict=config.ref_dict\n",
    "batch_dict=config.batch_dict\n",
    "batch_namedict=config.batch_namedict\n",
    "scriptdir='/igm/home/jbg001/git/germfilt/germline_variant_filtering/kf/script'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### BUILD DICTIONARIES TO RUN ALL\n",
    "### ALL WE NEED TO DO IS SPECIFY BATCHNUM AND GIABNUM\n",
    "## DATASETS.\n",
    "## IGM 1, IGM 2.\n",
    "imp.reload(f_nav)\n",
    "prepydict={0:'noprepy',1:'prepy'}\n",
    "rundict={1:[1,3,4],\n",
    "           2:[1,2],\n",
    "           3:[3,4,5],\n",
    "           9:[3,4],\n",
    "         \n",
    "           5:[3,4],\n",
    "           12:[3,4],\n",
    "           7:[3,4],\n",
    "           10:[6,7]\n",
    "        }\n",
    "\n",
    "prep_metadict={batchnum:{giabnum:{} for giabnum in rundict[batchnum]} for batchnum in rundict}\n",
    "hapdict={}\n",
    "for batchnum in rundict:\n",
    "#for batchnum in [12]:\n",
    "    for giabnum in rundict[batchnum]:\n",
    "        for prepyflag in [0]:\n",
    "            ## PREPROCESS\n",
    "            sig=str(batchnum)+'_'+str(giabnum)+'_'+str(prepyflag)\n",
    "            run_identifier='03062020_noon_'+prepydict[prepyflag]\n",
    "            preprocessing_dict,happy_summaryfile=f_nav.init_preprocess(batchnum,giabnum,run_identifier,prepyflag)\n",
    "            prep_metadict[batchnum][giabnum][prepyflag]=preprocessing_dict.copy()\n",
    "            preprofile=preprocessing_dict['vcf_query_preprocessed']\n",
    "            filtfile=preprofile.replace('preprocessed','filtered')\n",
    "            if os.path.exists(happy_summaryfile):\n",
    "                #print(\"Found \"+sig)\n",
    "                #print(os.path.getsize(preprofile),\n",
    "                #      os.path.getsize(filtfile))\n",
    "\n",
    "                hapdict[sig]=pd.read_csv(happy_summaryfile).assign(sig=sig)\n",
    "                xx=2\n",
    "            else:\n",
    "                xx=2\n",
    "                print(\"Missed \"+sig)\n",
    "                #print(happy_summaryfile)\n",
    "os.system('rm /igm/home/jbg001/git/germfilt/germline_variant_filtering/kf/script/run_*preprocess*')\n",
    "#prep_metadict[10][6][0]['vcf_query_preprocessed']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## BUILD TRAINING COMBO DICT\n",
    "combodict={1:{1:[1],2:[3],3:[4],4:[3,4]},\n",
    "           2:{1:[1],2:[2]},\n",
    "           3:{1:[3],2:[4],3:[3,4],4:[5]},\n",
    "           9:{1:[3],2:[4]},\n",
    "           \n",
    "           5:{1:[3],2:[4]},\n",
    "           12:{1:[3],2:[4]},\n",
    "           7:{1:[3],2:[4]},\n",
    "           10:{1:[6],2:[7]}\n",
    "          }\n",
    "           "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/igm/home/jbg001/miniconda3/envs/rnaenv/lib/python3.6/site-packages/ipykernel_launcher.py:28: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "/igm/home/jbg001/miniconda3/envs/rnaenv/lib/python3.6/site-packages/ipykernel_launcher.py:29: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## TRAIN \n",
    "imp.reload(f_nav)\n",
    "perfdict={}\n",
    "featimpdict={}\n",
    "train_metadict={batchnum:{comboind:{prepyflag:{} for prepyflag in [0,1]} for comboind in combodict[batchnum]} for batchnum in combodict}\n",
    "train_session='granular_train_032020_afternoon'\n",
    "mod_paramdict=f_nav.grab_mod_paramdict_generic_granular('X')\n",
    "\n",
    "for batchnum in combodict:\n",
    "    for combonum in combodict[batchnum]:\n",
    "        for prepyflag in [0]:\n",
    "            sig='_'.join([str(x) for x in [batchnum,combonum,prepyflag]])\n",
    "            xxx=3\n",
    "            train_paramdict=f_nav.init_train_multi(train_session,batchnum,combonum,prepyflag,mod_paramdict,prep_metadict,combodict)\n",
    "            train_metadict[batchnum][combonum][prepyflag]=train_paramdict.copy()\n",
    "            try:\n",
    "                perffile=train_paramdict['perffile']\n",
    "                perfdict[sig]=pd.read_csv(perffile).assign(sig=sig)\n",
    "                featimpdict[sig]=pd.read_csv(train_paramdict['perffile'].replace('allmod_summary','feature_importances')).assign(sig=sig)\n",
    "                #print(\"Found \"+sig)\n",
    "                #print(\"exists \"+perffile)\n",
    "            except:\n",
    "                print(\"Missing \"+sig)\n",
    "                #print(train_paramdict['traindir'])\n",
    "                xx=2\n",
    "#os.system('rm /igm/home/jbg001/git/germfilt/germline_variant_filtering/kf/script/run_train*')\n",
    "try:\n",
    "    perfdat=pd.concat(perfdict,ignore_index=True)\n",
    "    featdat=pd.concat(featimpdict,ignore_index=True)\n",
    "except:\n",
    "    xx=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## EXAMINE RESULTS, SET THRESH\n",
    "if 1==1:\n",
    "    perfdat['batch']=perfdat['sig'].apply(lambda x: int(x.split('_')[0]))\n",
    "    perfdat['FP_base']=perfdat['FP']+perfdat['TN']\n",
    "    perfdat['TP_base']=perfdat['TP']+perfdat['FN']\n",
    "    sig='5_1_0'\n",
    "# 5,7,10\n",
    "viewcols=['vartype','zyg','thresh','n','TP','FP','FN','TN','sensitivity','specificity','prec_base']\n",
    "#perfdat[(perfdat.batchnum==7)\n",
    "#       & (perfdat.sig==sig)][viewcols]\n",
    "#perfdat[(perfdat.sig==sig) & (perfdat.zyg=='hom')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "tabdat=pd.read_csv(train_metadict[5][1][0]['perffile'])\n",
    "tabdat['modfile']='moddir/mod_'+tabdat.vartype+'_'+tabdat.zyg+'_'+tabdat.modid.apply(str)+'/mod_0.pkl'\n",
    "keepcols=['vartype','zyg','modid','thresh','n','TP','FP','FN','TN']\n",
    "taboutdat=pd.concat([tabdat[tabdat.thresh>=.1].sort_values(['thresh','modid']).head(15),\n",
    "tabdat[tabdat.thresh>=.1].sort_values(['thresh','modid']).tail(1)],ignore_index=True)[keepcols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "| vartype | zyg | modid | thresh | n | TP | FP | FN | TN | \n",
      "| --- | --- | --- | --- | --- | --- | --- | --- | --- |  \n",
      "| SNP | het | 0 | 0.1 | 456631 | 454229 | 1166 | 80 | 1156 |  \n",
      "| SNP | hom | 1 | 0.1 | 295853 | 295824 | 29 | 0 | 0 |  \n",
      "| I1_5 | het | 2 | 0.1 | 27187 | 26409 | 453 | 9 | 316 |  \n",
      "| I1_5 | hom | 3 | 0.1 | 19533 | 19102 | 353 | 7 | 71 |  \n",
      "| I6_15 | het | 4 | 0.1 | 2369 | 2286 | 59 | 1 | 23 |  \n",
      "| I6_15 | hom | 5 | 0.1 | 1420 | 1306 | 73 | 13 | 28 |  \n",
      "| I16_PLUS | het | 6 | 0.1 | 493 | 461 | 11 | 0 | 21 |  \n",
      "| I16_PLUS | hom | 7 | 0.1 | 281 | 271 | 5 | 1 | 4 |  \n",
      "| D1_5 | het | 8 | 0.1 | 31999 | 30306 | 834 | 31 | 828 |  \n",
      "| D1_5 | hom | 9 | 0.1 | 19461 | 19077 | 323 | 5 | 56 |  \n",
      "| D6_15 | het | 10 | 0.1 | 2842 | 2661 | 114 | 10 | 57 |  \n",
      "| D6_15 | hom | 11 | 0.1 | 1544 | 1445 | 77 | 7 | 15 |  \n",
      "| D16_PLUS | het | 12 | 0.1 | 655 | 604 | 21 | 7 | 23 |  \n",
      "| D16_PLUS | hom | 13 | 0.1 | 299 | 288 | 7 | 1 | 3 |  \n",
      "| SNP | het | 0 | 0.2 | 456631 | 454174 | 948 | 135 | 1374 |  \n",
      "| D16_PLUS | hom | 13 | 0.9 | 299 | 280 | 3 | 9 | 7 |  \n",
      "\n"
     ]
    }
   ],
   "source": [
    "def df_to_bitbucket_table(df):\n",
    "    outstring='| '\n",
    "    for j in df.columns:\n",
    "        outstring+=j+' | '\n",
    "    outstring+='\\n'\n",
    "    outstring+='| '\n",
    "    for j in df.columns:\n",
    "        outstring+='--- | '\n",
    "    outstring +=' \\n'\n",
    "    for i in range(len(df)):\n",
    "        outstring+='| '\n",
    "        for j in df.columns:\n",
    "            outstring+=str(df.iloc[i][j])+' | '\n",
    "        outstring+=' \\n'\n",
    "    return(outstring)\n",
    "outstring=df_to_bitbucket_table(taboutdat)\n",
    "print(outstring)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 True\n",
      "2 True\n",
      "3 True\n",
      "9 True\n",
      "5 True\n",
      "12 True\n",
      "7 True\n",
      "10 True\n"
     ]
    }
   ],
   "source": [
    "## APPLY\n",
    "imp.reload(f_nav)\n",
    "apply_dict={\n",
    " 1:{'train':'1_2_0','test':'1_3_0','thresh':{'SNP_het':.1,'D1_5_het':.1}},\n",
    " 2:{'train':'2_1_0','test':'2_2_0','thresh':{'SNP_het':.1,'D1_5_het':.1}},\n",
    " 3:{'train':'3_4_0','test':'3_2_0','thresh':{'SNP_het':.1,'D1_5_het':.05}},\n",
    " 9:{'train':'9_1_0','test':'9_2_0','thresh':{'SNP_het':.1,'D1_5_het':.1}},\n",
    "\n",
    " 5:{'train':'5_1_0','test':'5_2_0','thresh':{'SNP_het':.1,'D1_5_het':.1}},\n",
    " 12:{'train':'12_1_0','test':'12_2_0','thresh':{'SNP_het':.1,'D1_5_het':.1}},\n",
    " 7:{'train':'7_1_0','test':'7_2_0','thresh':{'SNP_het':.1,'D1_5_het':.1}},\n",
    " 10:{'train':'10_1_0','test':'10_2_0','thresh':{'SNP_het':.1,'D1_5_het':.1}}  \n",
    "}\n",
    "\n",
    "\n",
    "intdict={}\n",
    "extdict={}\n",
    "chrdict={x:'21' for x in apply_dict}\n",
    "testdict={}; lc=0\n",
    "for i in apply_dict:\n",
    "#for i in [5,12,7,10]:\n",
    "    trainsig=apply_dict[i]['train']\n",
    "    testsig=apply_dict[i]['test']\n",
    "    threshdict=apply_dict[i]['thresh']\n",
    "    batchnum=int(trainsig.split('_')[0])\n",
    "    combonum_train=int(trainsig.split('_')[1])\n",
    "    combonum_test=int(testsig.split('_')[1])\n",
    "    giabnum_train=combodict[batchnum][combonum_train][0]\n",
    "    giabnum_test=combodict[batchnum][combonum_test][0]\n",
    "    train_paramdict=train_metadict[batchnum][combonum_train][0]\n",
    "    mod_paramdict=f.jread(train_paramdict['mod_paramdict'])\n",
    "    for modnum in mod_paramdict:\n",
    "        modsig=mod_paramdict[modnum]['vartype']+'_'+mod_paramdict[modnum]['zyg']\n",
    "        if modsig in threshdict:\n",
    "            mod_paramdict[modnum]['thresh']=threshdict[modsig]\n",
    "        else:\n",
    "            mod_paramdict[modnum]['thresh']=0\n",
    "    preprocessing_dict=prep_metadict[batchnum][giabnum_test][0]\n",
    "    moddir_base=train_paramdict['traindir']+'/mod'\n",
    "    infile=preprocessing_dict['vcf_query_preprocessed']\n",
    "    locdir=preprocessing_dict['locdir']\n",
    "    sumfile_ext=locdir+'/summary_external.csv'\n",
    "    sumfile_int=locdir+'/summary_internal.csv'\n",
    "    truthfile=locdir+'/truth_filtered_ph.vcf.gz'\n",
    "    bedfile=locdir+'/giab_intersected_ph.bed'\n",
    "    refnum=batch_dict[batchnum]['assembly']\n",
    "    reffile=ref_dict[refnum]\n",
    "    filtflag=0\n",
    "    try:\n",
    "        intdict[i]=pd.read_csv(sumfile_int).assign(batch=i)\n",
    "    except:\n",
    "        xx=1\n",
    "    try:\n",
    "        extdict[i]=pd.read_csv(sumfile_ext).assign(batch=i)\n",
    "    except:\n",
    "        xx=1\n",
    "    if filtflag==1:\n",
    "        scriptfile=scriptdir+'/run_bedfilt_'+str(i)+'.sh'\n",
    "        chrtofilt=chrdict[i]\n",
    "        filtfile=infile.replace('.vcf','_filt_'+chrtofilt+'.vcf')\n",
    "        hapfile_old=locdir+'/happyout_ph.vcf.gz'\n",
    "        hapfile_new=locdir+'/happyout_ph_filt_'+chrtofilt+'.vcf.gz'\n",
    "        f_nav.run_bedfilt(infile,filtfile,chrtofilt,scriptfile)\n",
    "        outfile=locdir+'/applied_'+train_session+'_applyid_'+str(i)+'_filt_'+chrtofilt+'.vcf'\n",
    "        infile=filtfile\n",
    "        scriptfile=scriptdir+'/run_hapfilt_'+str(i)+'.sh'\n",
    "        truthfile_filt=truthfile.replace('.vcf.gz','_filt_'+chrtofilt+'.vcf.gz')\n",
    "        f_nav.run_bedfilt(truthfile,truthfile_filt,chrtofilt,scriptfile)\n",
    "        f_nav.run_bedfilt(hapfile_old,hapfile_new,chrtofilt,scriptfile)\n",
    "        truthfile=truthfile_filt\n",
    "    else:\n",
    "        outfile=locdir+'/autofiltered_'+train_session+'_appid_'+str(i)+'.vcf'\n",
    "    scriptfile=scriptdir+'/run_apply_'+str(i)+'.sh'\n",
    "    print(i,os.path.exists(sumfile_int))\n",
    "    #f_nav.run_apply_autofilt_papertesting(infile,outfile,moddir_base,modnumdict,mod_paramdict,thresh,scriptfile)\n",
    "    #f_nav.run_apply_autofilt_papertesting_simp(infile,outfile,mod_paramdict,thresh,scriptfile)\n",
    "    f_nav.run_apply_autofilt_papertesting_simp_threshindict(infile,outfile,mod_paramdict,scriptfile)\n",
    "    outstring=outfile.replace('.vcf','_happed')\n",
    "    scriptfile=scriptdir+'/run_happy_'+str(i)+'.sh'\n",
    "    f_nav.run_justhappy(outfile,truthfile,reffile,bedfile,outstring,scriptfile)\n",
    "    hapfile_extended=outstring+'.extended.csv'\n",
    "    try:\n",
    "        testdict[lc]=pd.read_csv(hapfile_extended).assign(\n",
    "         #   thresh=thresh,\n",
    "            batchnum=batchnum,giabnum_test=giabnum_test,giabnum_train=giabnum_train,appnum=i)\n",
    "        #print('Found '+testsig)\n",
    "    except:\n",
    "        xx=2\n",
    "        #print('Missing '+hapfile_extended)\n",
    "    lc+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "idat=pd.concat(intdict,ignore_index=True)\n",
    "xdat=pd.concat(extdict,ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "#grab_tpfp(idat[(idat.batch==3) & (idat.vartype!= 'SNP') & (idat.thresh==.05)].copy())\n",
    "imp.reload(config)\n",
    "batchnamedict=config.batch_namedict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "bigcutdict={}\n",
    "batch=1\n",
    "cutdict={\n",
    "    'SNP;het':.1,\n",
    "    'D1_5;het':.1}\n",
    "codes_to_filt=[]\n",
    "bigcutdict[batch]={'cuts':cutdict.copy(), 'codes':codes_to_filt.copy()}\n",
    "batch=12\n",
    "cutdict={\n",
    "    'SNP;het':.4,\n",
    "    'I6_15;het':.3,\n",
    "    'I1_5;het':.4,\n",
    "    'I16_PLUS;het':.3,   \n",
    "    'D6_15;het':.3,\n",
    "    'D1_5;het':.3}\n",
    "\n",
    "batch=3\n",
    "cutdict={\n",
    "    'SNP;het':.4,\n",
    "    'D1_5;het':.05}\n",
    "codes_to_filt=['gatkRecommendIndelFilter','VQSRTrancheSNP99.9']\n",
    "bigcutdict[batch]={'cuts':cutdict.copy(), 'codes':codes_to_filt.copy()}\n",
    "batch=12\n",
    "cutdict={\n",
    "    'SNP;het':.4,\n",
    "    'I6_15;het':.3,\n",
    "    'I1_5;het':.4,\n",
    "    'I16_PLUS;het':.3,   \n",
    "    'D6_15;het':.3,\n",
    "    'D1_5;het':.3}\n",
    "codes_to_filt=['DRAGENHardQUAL','LowDepth']\n",
    "bigcutdict[batch]={'cuts':cutdict.copy(), 'codes':codes_to_filt.copy()}\n",
    "batch=10\n",
    "cutdict={\n",
    "    'SNP;het':.2,\n",
    "    'I6_15;het':.1,\n",
    "    'I1_5;het':.1,\n",
    "    'I16_PLUS;het':.5,   \n",
    "    'D6_15;het':.1,\n",
    "    'D1_5;het':.1}\n",
    "codes_to_filt=['VQLOW']\n",
    "bigcutdict[batch]={'cuts':cutdict.copy(), 'codes':codes_to_filt.copy()}\n",
    "batch=5\n",
    "cutdict={\n",
    "    'SNP;het':.4,\n",
    "    'I6_15;het':.1,\n",
    "    'I1_5;het':.1,\n",
    "    'I16_PLUS;het':.1,   \n",
    "    'D6_15;het':.1,\n",
    "    'D1_5;het':.1,\n",
    "    'I16_PLUS;het':.1,\n",
    "    'D1_5;hom':.1\n",
    "}\n",
    "codes_to_filt=[]\n",
    "bigcutdict[batch]={'cuts':cutdict.copy(), 'codes':codes_to_filt.copy()}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 /igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_1_giab_1/query_preprocessed_ph.vcf.gz\n",
      "/igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_1_giab_1/query_preprocessed_ph.vcf.gz\n",
      "3 /igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_3_giab_3/query_preprocessed_ph.vcf.gz\n",
      "/igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_3_giab_3/query_preprocessed_ph.vcf.gz\n",
      "12 /igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_12_giab_3/query_preprocessed_ph.vcf.gz\n",
      "/igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_12_giab_3/query_preprocessed_ph.vcf.gz\n",
      "10 /igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_10_giab_6/query_preprocessed_ph.vcf.gz\n",
      "/igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_10_giab_6/query_preprocessed_ph.vcf.gz\n",
      "5 /igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_5_giab_3/query_preprocessed_ph.vcf.gz\n",
      "/igm/projects/FOR_JEFF/germfilt/tm/03062020_noon_noprepy_batch_5_giab_3/query_preprocessed_ph.vcf.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/igm/home/jbg001/miniconda3/envs/rnaenv/lib/python3.6/site-packages/ipykernel_launcher.py:13: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  del sys.path[0]\n"
     ]
    }
   ],
   "source": [
    "sumdict={}\n",
    "for batch in bigcutdict:\n",
    "#for batch in [3,10]:\n",
    "    cutdict=bigcutdict[batch]['cuts']\n",
    "    codes_to_filt=bigcutdict[batch]['codes']\n",
    "    giabnum_train=combodict[batch][int(apply_dict[batch]['train'].split('_')[1])][0]\n",
    "    giabnum_test=combodict[batch][int(apply_dict[batch]['test'].split('_')[1])][0]\n",
    "    vcffile=prep_metadict[batch][list(prep_metadict[batch].keys())[0]][0]['vcf_query_preprocessed']\n",
    "    print(batch,vcffile)\n",
    "    print(vcffile)\n",
    "    locdat=idat[idat.batch==batch].copy()\n",
    "    zdat=locdat[locdat.thresh==0.1].groupby(['batch','vartype','zyg','y'])['n'].sum().reset_index().assign(ypred=1,thresh=0)\n",
    "    locdat=pd.concat([locdat,zdat])\n",
    "    locdat['vz']=locdat['vartype']+';'+locdat['zyg']\n",
    "    for x in locdat.vz.unique():\n",
    "        if x not in cutdict:\n",
    "            cutdict[x]=0\n",
    "    zdat=locdat[locdat.thresh==0.1].groupby(['batch','vartype','zyg','y'])['n'].sum().reset_index().assign(ypred=1,thresh=0)\n",
    "    locdat['thresh_chosen']=locdat['vz'].apply(lambda x: cutdict[x])\n",
    "    hokdat=locdat[locdat.thresh==locdat.thresh_chosen].drop(columns=['thresh_chosen','vz'])\n",
    "    hokdat['vartype_full']='INDEL'\n",
    "    hokdat.loc[hokdat.vartype=='SNP','vartype_full']='SNP'\n",
    "    groupcols=['batch','vartype_full','y','ypred']\n",
    "    intdat=hokdat.groupby(groupcols)['n'].sum().reset_index().assign(filt='int')\n",
    "    exdat=xdat[xdat.batch==batch].copy()\n",
    "    exdat['ypred']=1\n",
    "    for code_to_filt in codes_to_filt:\n",
    "        exdat.loc[exdat.FILT_string.str.contains(code_to_filt),'ypred']=0\n",
    "    exdat['vartype_full']='INDEL'\n",
    "    exdat.loc[exdat.vartype=='SNP','vartype_full']='SNP'\n",
    "    extdat=exdat.groupby(groupcols)['n'].sum().reset_index().assign(filt='ext')\n",
    "    exsumdat=extdat.groupby(['vartype_full','y','ypred'])['n'].sum().reset_index()\n",
    "    intsumdat=intdat.groupby(['vartype_full','y','ypred'])['n'].sum().reset_index()\n",
    "    sumdat=pd.concat([exsumdat.assign(filtstyle='ext'),intsumdat.assign(filtstyle='int')]).assign(batch=batch,\n",
    "                                                  giabnum_train=giabnum_train,\n",
    "                                                  giabnum_test=giabnum_test)\n",
    "    sumdict[batch]=sumdat.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "def grab_tpfp(df):\n",
    "    tosscols=['y','ypred','n']\n",
    "    groupcols=[x for x in df.columns if x not in tosscols+['TP','FP','FN','TN']]\n",
    "    df['TP']=df['y']*df['ypred']*df['n']\n",
    "    df['FP']=(1-df['y'])*df['ypred']*df['n']\n",
    "    df['FN']=df['y']*(1-df['ypred'])*df['n']\n",
    "    df['TN']=(1-df['y'])*(1-df['ypred'])*df['n']\n",
    "    gf=df.groupby(groupcols).agg({x:'sum' for x in [y+z for y in ['T','F'] for z in ['P','N']]}).reset_index()\n",
    "    return(gf)\n",
    "humdat=pd.concat(sumdict,ignore_index=True).rename(columns={'vartype_full':'vartype'})\n",
    "sumdat=grab_tpfp(humdat)\n",
    "sumdat['recall']=sumdat['TP']*1.0/(sumdat.TP+sumdat.FN)\n",
    "sumdat['precision']=sumdat['TP']*1.0/(sumdat.TP+sumdat.FP)\n",
    "sumdat['specificity']=sumdat['TN']*1.0/(sumdat.TN+sumdat.FP)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "imp.reload(config)\n",
    "batch_namedict=config.batch_namedict\n",
    "sumdat['batchname']=sumdat['batch'].apply(lambda x: batch_namedict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vartype</th>\n",
       "      <th>batch</th>\n",
       "      <th>batchname</th>\n",
       "      <th>giabnum_train</th>\n",
       "      <th>giabnum_test</th>\n",
       "      <th>TP</th>\n",
       "      <th>FP</th>\n",
       "      <th>FN</th>\n",
       "      <th>TN</th>\n",
       "      <th>filter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>IGM GATK 4.1.2 WES</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>40939.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>1</td>\n",
       "      <td>IGM GATK 4.1.2 WES</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3210.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SNP</td>\n",
       "      <td>3</td>\n",
       "      <td>Oslo joint and single exome</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>50456.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SNP</td>\n",
       "      <td>3</td>\n",
       "      <td>Oslo joint and single exome</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>50431.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>VQSRTrancheSNP99.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>3</td>\n",
       "      <td>Oslo joint and single exome</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4823.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>267.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>3</td>\n",
       "      <td>Oslo joint and single exome</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4823.0</td>\n",
       "      <td>369.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>gatkRecommendIndelFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SNP</td>\n",
       "      <td>5</td>\n",
       "      <td>IGM NICU WGS</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3028517.0</td>\n",
       "      <td>1442.0</td>\n",
       "      <td>547.0</td>\n",
       "      <td>3008.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>5</td>\n",
       "      <td>IGM NICU WGS</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>425104.0</td>\n",
       "      <td>8737.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>2550.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SNP</td>\n",
       "      <td>10</td>\n",
       "      <td>Complete Genomics WGS</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>2897506.0</td>\n",
       "      <td>8453.0</td>\n",
       "      <td>993.0</td>\n",
       "      <td>19510.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SNP</td>\n",
       "      <td>10</td>\n",
       "      <td>Complete Genomics WGS</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>2889866.0</td>\n",
       "      <td>11614.0</td>\n",
       "      <td>8633.0</td>\n",
       "      <td>16349.0</td>\n",
       "      <td>VQLOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>10</td>\n",
       "      <td>Complete Genomics WGS</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>281227.0</td>\n",
       "      <td>9669.0</td>\n",
       "      <td>727.0</td>\n",
       "      <td>99466.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>10</td>\n",
       "      <td>Complete Genomics WGS</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>268673.0</td>\n",
       "      <td>32219.0</td>\n",
       "      <td>13281.0</td>\n",
       "      <td>76916.0</td>\n",
       "      <td>VQLOW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SNP</td>\n",
       "      <td>12</td>\n",
       "      <td>IGM DRAGEN WGS Exp. 3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>1768085.0</td>\n",
       "      <td>691725.0</td>\n",
       "      <td>207770.0</td>\n",
       "      <td>353614.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SNP</td>\n",
       "      <td>12</td>\n",
       "      <td>IGM DRAGEN WGS Exp. 3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>1973536.0</td>\n",
       "      <td>1042420.0</td>\n",
       "      <td>2319.0</td>\n",
       "      <td>2919.0</td>\n",
       "      <td>DRAGENHardQUAL,LowDepth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>12</td>\n",
       "      <td>IGM DRAGEN WGS Exp. 3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>233467.0</td>\n",
       "      <td>122300.0</td>\n",
       "      <td>21964.0</td>\n",
       "      <td>55118.0</td>\n",
       "      <td>AutoFilter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>12</td>\n",
       "      <td>IGM DRAGEN WGS Exp. 3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>255154.0</td>\n",
       "      <td>176062.0</td>\n",
       "      <td>277.0</td>\n",
       "      <td>1356.0</td>\n",
       "      <td>DRAGENHardQUAL,LowDepth</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   vartype  batch                    batchname  giabnum_train  giabnum_test  \\\n",
       "13     SNP      1           IGM GATK 4.1.2 WES              3             4   \n",
       "4    INDEL      1           IGM GATK 4.1.2 WES              3             4   \n",
       "14     SNP      3  Oslo joint and single exome              5             4   \n",
       "9      SNP      3  Oslo joint and single exome              5             4   \n",
       "5    INDEL      3  Oslo joint and single exome              5             4   \n",
       "0    INDEL      3  Oslo joint and single exome              5             4   \n",
       "15     SNP      5                 IGM NICU WGS              3             4   \n",
       "6    INDEL      5                 IGM NICU WGS              3             4   \n",
       "16     SNP     10        Complete Genomics WGS              6             7   \n",
       "11     SNP     10        Complete Genomics WGS              6             7   \n",
       "7    INDEL     10        Complete Genomics WGS              6             7   \n",
       "2    INDEL     10        Complete Genomics WGS              6             7   \n",
       "17     SNP     12        IGM DRAGEN WGS Exp. 3              3             4   \n",
       "12     SNP     12        IGM DRAGEN WGS Exp. 3              3             4   \n",
       "8    INDEL     12        IGM DRAGEN WGS Exp. 3              3             4   \n",
       "3    INDEL     12        IGM DRAGEN WGS Exp. 3              3             4   \n",
       "\n",
       "           TP         FP        FN        TN                    filter  \n",
       "13    40939.0       63.0       8.0     266.0                AutoFilter  \n",
       "4      3210.0       80.0       9.0     119.0                AutoFilter  \n",
       "14    50456.0       49.0       9.0      85.0                AutoFilter  \n",
       "9     50431.0       77.0      34.0      57.0        VQSRTrancheSNP99.9  \n",
       "5      4823.0      271.0       1.0     267.0                AutoFilter  \n",
       "0      4823.0      369.0       1.0     169.0  gatkRecommendIndelFilter  \n",
       "15  3028517.0     1442.0     547.0    3008.0                AutoFilter  \n",
       "6    425104.0     8737.0     169.0    2550.0                AutoFilter  \n",
       "16  2897506.0     8453.0     993.0   19510.0                AutoFilter  \n",
       "11  2889866.0    11614.0    8633.0   16349.0                     VQLOW  \n",
       "7    281227.0     9669.0     727.0   99466.0                AutoFilter  \n",
       "2    268673.0    32219.0   13281.0   76916.0                     VQLOW  \n",
       "17  1768085.0   691725.0  207770.0  353614.0                AutoFilter  \n",
       "12  1973536.0  1042420.0    2319.0    2919.0   DRAGENHardQUAL,LowDepth  \n",
       "8    233467.0   122300.0   21964.0   55118.0                AutoFilter  \n",
       "3    255154.0   176062.0     277.0    1356.0   DRAGENHardQUAL,LowDepth  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sumdat_mid=sumdat.sort_values(['batch','vartype','filtstyle'],ascending=[True,False,False]).drop(columns=['recall','precision','specificity'])\n",
    "sumdat_mid['filter']=sumdat_mid.apply(lambda x: \n",
    "    'AutoFilter' if x['filtstyle']=='int' \n",
    "    else ','.join(bigcutdict[x['batch']]['codes']),axis=1)\n",
    "sumdat_mid['filter']=sumdat_mid.apply(lambda x: \n",
    "    x['filter'].split(',')[0] if (x['vartype']=='INDEL') and ('IndelFilter' in x['filter']) else\n",
    "    (x['filter'].split(',')[1] if (x['vartype']=='SNP') and ('IndelFilter' in x['filter']) else x['filter']),axis=1)\n",
    "                                  \n",
    "sumdat_mid=sumdat_mid[sumdat_mid.TN > 0 ]\n",
    "outcols=['vartype','batch','batchname','giabnum_train','giabnum_test','TP','FP','FN','TN','filter']\n",
    "sumdat_mid[outcols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sumdat' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-104-1749ace85259>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mhumdat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msumdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TP'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'FP'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'FN'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'TN'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;31m#.rename(columns={'vartype_full':'vartype'})\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mvalcols\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'recall'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'precision'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mpivcols\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'filtstyle'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mgroupcols\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mhumdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalcols\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mpivcols\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#pd.pivot_table(humdat,index=groupcols,values=valcols,columns=pivcols)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sumdat' is not defined"
     ]
    }
   ],
   "source": [
    "humdat=sumdat.drop(columns=['TP','FP','FN','TN'])#.rename(columns={'vartype_full':'vartype'})\n",
    "valcols=['recall','precision']\n",
    "pivcols=['filtstyle']\n",
    "groupcols=[x for x in humdat.columns if x not in valcols+pivcols]\n",
    "#pd.pivot_table(humdat,index=groupcols,values=valcols,columns=pivcols)\n",
    "wumdat=humdat[(humdat.batch!= 12)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'wumdat' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-105-21110b159581>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mvartype\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'SNP'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'INDEL'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mfiltstyle\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'int'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'ext'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m             \u001b[0mplotdat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mwumdat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwumdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbatch\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mbatch\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mwumdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfiltstyle\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mfiltstyle\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mwumdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvartype\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mvartype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m             \u001b[0mfiltname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'AutoFilter'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mfiltstyle\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'ext'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'wumdat' is not defined"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAD8CAYAAAB0IB+mAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAADU9JREFUeJzt3GGI5Hd9x/H3xztTaYym9FaQu9Ok9NJ42ELSJU0Raoq2XPLg7oFF7iBYJXhgGylVhBRLlPjIhloQrtWTilXQGH0gC57cA40ExAu3ITV4FyLb03oXhawxzZOgMe23D2bSna53mX92Z3cv+32/4GD+//ntzJcfe++dndmZVBWSpO3vFVs9gCRpcxh8SWrC4EtSEwZfkpow+JLUhMGXpCamBj/JZ5M8meT7l7g+ST6ZZCnJo0lunP2YkqT1GvII/3PAgRe5/lZg3/jfUeBf1j+WJGnWpga/qh4Efv4iSw4Bn6+RU8DVSV4/qwElSbOxcwa3sRs4P3F8YXzup6sXJjnK6LcArrzyyj+8/vrrZ3D3ktTHww8//LOqmlvL184i+INV1XHgOMD8/HwtLi5u5t1L0stekv9c69fO4q90ngD2ThzvGZ+TJF1GZhH8BeBd47/WuRl4pqp+7ekcSdLWmvqUTpIvAbcAu5JcAD4CvBKgqj4FnABuA5aAZ4H3bNSwkqS1mxr8qjoy5foC/npmE0mSNoTvtJWkJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJamJQcFPciDJ40mWktx1kevfkOSBJI8keTTJbbMfVZK0HlODn2QHcAy4FdgPHEmyf9Wyvwfur6obgMPAP896UEnS+gx5hH8TsFRV56rqOeA+4NCqNQW8Znz5tcBPZjeiJGkWhgR/N3B+4vjC+NykjwK3J7kAnADef7EbSnI0yWKSxeXl5TWMK0laq1m9aHsE+FxV7QFuA76Q5Nduu6qOV9V8Vc3Pzc3N6K4lSUMMCf4TwN6J4z3jc5PuAO4HqKrvAq8Cds1iQEnSbAwJ/mlgX5Jrk1zB6EXZhVVrfgy8DSDJmxgF3+dsJOkyMjX4VfU8cCdwEniM0V/jnElyT5KD42UfBN6b5HvAl4B3V1Vt1NCSpJdu55BFVXWC0Yuxk+funrh8FnjLbEeTJM2S77SVpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDUxKPhJDiR5PMlSkrsuseadSc4mOZPki7MdU5K0XjunLUiyAzgG/BlwATidZKGqzk6s2Qf8HfCWqno6yes2amBJ0toMeYR/E7BUVeeq6jngPuDQqjXvBY5V1dMAVfXkbMeUJK3XkODvBs5PHF8Yn5t0HXBdku8kOZXkwMVuKMnRJItJFpeXl9c2sSRpTWb1ou1OYB9wC3AE+EySq1cvqqrjVTVfVfNzc3MzumtJ0hBDgv8EsHfieM/43KQLwEJV/aqqfgj8gNEPAEnSZWJI8E8D+5Jcm+QK4DCwsGrN1xg9uifJLkZP8Zyb4ZySpHWaGvyqeh64EzgJPAbcX1VnktyT5OB42UngqSRngQeAD1XVUxs1tCTppUtVbckdz8/P1+Li4pbctyS9XCV5uKrm1/K1vtNWkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgYFP8mBJI8nWUpy14use0eSSjI/uxElSbMwNfhJdgDHgFuB/cCRJPsvsu4q4G+Ah2Y9pCRp/YY8wr8JWKqqc1X1HHAfcOgi6z4GfBz4xQznkyTNyJDg7wbOTxxfGJ/7P0luBPZW1ddf7IaSHE2ymGRxeXn5JQ8rSVq7db9om+QVwCeAD05bW1XHq2q+qubn5ubWe9eSpJdgSPCfAPZOHO8Zn3vBVcCbgW8n+RFwM7DgC7eSdHkZEvzTwL4k1ya5AjgMLLxwZVU9U1W7quqaqroGOAUcrKrFDZlYkrQmU4NfVc8DdwIngceA+6vqTJJ7khzc6AElSbOxc8iiqjoBnFh17u5LrL1l/WNJkmbNd9pKUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpoYFPwkB5I8nmQpyV0Xuf4DSc4meTTJN5O8cfajSpLWY2rwk+wAjgG3AvuBI0n2r1r2CDBfVX8AfBX4h1kPKklanyGP8G8ClqrqXFU9B9wHHJpcUFUPVNWz48NTwJ7ZjilJWq8hwd8NnJ84vjA+dyl3AN+42BVJjiZZTLK4vLw8fEpJ0rrN9EXbJLcD88C9F7u+qo5X1XxVzc/Nzc3yriVJU+wcsOYJYO/E8Z7xuf8nyduBDwNvrapfzmY8SdKsDHmEfxrYl+TaJFcAh4GFyQVJbgA+DRysqidnP6Ykab2mBr+qngfuBE4CjwH3V9WZJPckOThedi/wauArSf49ycIlbk6StEWGPKVDVZ0ATqw6d/fE5bfPeC5J0oz5TltJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaGBT8JAeSPJ5kKcldF7n+N5J8eXz9Q0mumfWgkqT1mRr8JDuAY8CtwH7gSJL9q5bdATxdVb8L/BPw8VkPKklanyGP8G8ClqrqXFU9B9wHHFq15hDwb+PLXwXeliSzG1OStF47B6zZDZyfOL4A/NGl1lTV80meAX4b+NnkoiRHgaPjw18m+f5aht6GdrFqrxpzL1a4FyvcixW/t9YvHBL8mamq48BxgCSLVTW/mfd/uXIvVrgXK9yLFe7FiiSLa/3aIU/pPAHsnTjeMz530TVJdgKvBZ5a61CSpNkbEvzTwL4k1ya5AjgMLKxaswD85fjyXwDfqqqa3ZiSpPWa+pTO+Dn5O4GTwA7gs1V1Jsk9wGJVLQD/CnwhyRLwc0Y/FKY5vo65txv3YoV7scK9WOFerFjzXsQH4pLUg++0laQmDL4kNbHhwfdjGVYM2IsPJDmb5NEk30zyxq2YczNM24uJde9IUkm27Z/kDdmLJO8cf2+cSfLFzZ5xswz4P/KGJA8keWT8/+S2rZhzoyX5bJInL/VepYx8crxPjya5cdANV9WG/WP0Iu9/AL8DXAF8D9i/as1fAZ8aXz4MfHkjZ9qqfwP34k+B3xxffl/nvRivuwp4EDgFzG/13Fv4fbEPeAT4rfHx67Z67i3ci+PA+8aX9wM/2uq5N2gv/gS4Efj+Ja6/DfgGEOBm4KEht7vRj/D9WIYVU/eiqh6oqmfHh6cYvedhOxryfQHwMUafy/SLzRxukw3Zi/cCx6rqaYCqenKTZ9wsQ/aigNeML78W+MkmzrdpqupBRn/xeCmHgM/XyCng6iSvn3a7Gx38i30sw+5Lramq54EXPpZhuxmyF5PuYPQTfDuauhfjX1H3VtXXN3OwLTDk++I64Lok30lyKsmBTZtucw3Zi48Ctye5AJwA3r85o112XmpPgE3+aAUNk+R2YB5461bPshWSvAL4BPDuLR7lcrGT0dM6tzD6re/BJL9fVf+1pVNtjSPA56rqH5P8MaP3/7y5qv5nqwd7OdjoR/h+LMOKIXtBkrcDHwYOVtUvN2m2zTZtL64C3gx8O8mPGD1HubBNX7gd8n1xAVioql9V1Q+BHzD6AbDdDNmLO4D7Aarqu8CrGH2wWjeDerLaRgffj2VYMXUvktwAfJpR7Lfr87QwZS+q6pmq2lVV11TVNYxezzhYVWv+0KjL2JD/I19j9OieJLsYPcVzbjOH3CRD9uLHwNsAkryJUfCXN3XKy8MC8K7xX+vcDDxTVT+d9kUb+pRObdzHMrzsDNyLe4FXA18Zv27946o6uGVDb5CBe9HCwL04Cfx5krPAfwMfqqpt91vwwL34IPCZJH/L6AXcd2/HB4hJvsToh/yu8esVHwFeCVBVn2L0+sVtwBLwLPCeQbe7DfdKknQRvtNWkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJauJ/Acz2XLpusNoKAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import math\n",
    "fig,ax=plt.subplots()\n",
    "cols=plt.rcParams['axes.prop_cycle'].by_key()['color']\n",
    "plt.rc('axes', labelsize=12) \n",
    "markerdict={'ext':'^','int':'o'}\n",
    "lc=0\n",
    "#for batch in bigcutdict:\n",
    "batch_namedict_loc={3:'Oslo WES', 10:'Complete Genomics WGS'}\n",
    "external_filtdict={3:{'SNP':'VQSR','INDEL':'gatkRecommendedIndelFilter'},\n",
    "                  10:{x:'VQLOW' for x in ['SNP','INDEL']}}\n",
    "for batch in [3,10]:\n",
    "    for vartype in ['SNP','INDEL']:\n",
    "        for filtstyle in ['int','ext']:\n",
    "            plotdat=wumdat[(wumdat.batch==batch)  & (wumdat.filtstyle==filtstyle) & (wumdat.vartype==vartype)]\n",
    "            filtname='AutoFilter'\n",
    "            if filtstyle=='ext':\n",
    "                filtname=external_filtdict[batch][vartype]\n",
    "            cols_custom=[cols[2],cols[3],cols[0],cols[1]]\n",
    "            ax.scatter(x=plotdat.recall.apply(lambda x: -math.log10(1-x)),\n",
    "                       y=plotdat.precision.apply(lambda x: -math.log10(1-x)),\n",
    "                        #x=plotdat.precision,y=plotdat.recall,\n",
    "                       label=batch_namedict_loc[batch]+' '+vartype+'s; '+filtname,\n",
    "                       c=cols_custom[lc],marker=markerdict[filtstyle],s=200)\n",
    "        lc+=1\n",
    "plt.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\")\n",
    "xtick_labs=[1-(10 ** (-x)) for x in plt.xticks()[0]]\n",
    "ytick_labs=[1-(10 ** (-x)) for x in plt.xticks()[0]]\n",
    "xtick_labs=[round(x,4) for x in xtick_labs]\n",
    "ytick_labs=[round(x,4) for x in ytick_labs]\n",
    "\n",
    "plt.xticks(ticks=plt.xticks()[0],labels=xtick_labs)\n",
    "plt.yticks(ticks=plt.yticks()[0],labels=ytick_labs)\n",
    "plt.xlabel('Recall')\n",
    "plt.ylabel('Precision')\n",
    "\n",
    "#pfdir='/igm/home/jbg001/git/germfilt/germline_variant_filtering/plotfin'\n",
    "for x in ['pdf','png']:\n",
    "    plotfile=pfdir+'/precrec_forpaper.'+x\n",
    "    plt.savefig(plotfile,bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sumdat' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-106-a4f76fe86232>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m sumdat[(sumdat.batch==5)\n\u001b[0m\u001b[1;32m      2\u001b[0m     & (sumdat.vartype=='SNP')]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sumdat' is not defined"
     ]
    }
   ],
   "source": [
    "sumdat[(sumdat.batch==5)\n",
    "    & (sumdat.vartype=='SNP')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'wumdat' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-107-38a39efe8580>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mwumdat\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mwumdat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecall\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'wumdat' is not defined"
     ]
    }
   ],
   "source": [
    "wumdat[wumdat.recall==1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "5",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-108-c87e8cde852f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msumdict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 5"
     ]
    }
   ],
   "source": [
    "sumdict[5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "#xdat[['FILT_string','batch']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:rnaenv]",
   "language": "python",
   "name": "conda-env-rnaenv-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
